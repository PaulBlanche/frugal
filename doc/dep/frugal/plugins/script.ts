export * from 'frugal/plugins/script.ts';
