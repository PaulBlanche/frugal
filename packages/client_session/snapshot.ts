import * as murmur from '../murmur/mod.ts';

type Node = {
    name: string;
    attributes: { name: string; value: string }[];
    content: string;
    hash: string;
    element: Element;
};

export type Head = Node[];

export function serialize(head: HTMLHeadElement) {
    const snapshot: Head = [];

    head.childNodes.forEach((node) => {
        if (!(node instanceof Element)) {
            return;
        }

        const hash = new murmur.Hash();

        const name = node.nodeName.toLowerCase();
        const content = node.textContent ?? '';

        const sorted: string[] = [];
        const attributes: { name: string; value: string }[] = [];
        for (let i = 0, length = node.attributes.length; i < length; i++) {
            const attribute = node.attributes[i];
            attributes.push({
                name: attribute.name,
                value: attribute.value,
            });
            sorted.push(`${attribute.name}="${attribute.value}"`);
        }

        hash.update(name);

        sorted.sort();
        sorted.forEach((value) => {
            hash.update(value);
        });

        hash.update(content);

        snapshot.push({
            name,
            attributes,
            content,
            hash: hash.digest(),
            element: node,
        });
    });

    return snapshot;
}
