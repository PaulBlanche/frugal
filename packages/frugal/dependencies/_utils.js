/**
 * @template T
 * @param {AsyncIterable<T>} stream
 * @param {number} interval
 */
export function* debounce(stream, interval) {
    /** @type {boolean | undefined} */
    let first; // is this first event?  will pass
    /** @type {T[]} */
    let buffer = []; // the last event raised
    /** @type {Promise<T[]>} */
    let deferred; // deferred promise instance
    /** @type {(value: T[]) => void} */
    let resolve; // resolve method for deferred promise

    /** @param {boolean} isFirst */
    function reset(isFirst) {
        first = isFirst;
        buffer.length = 0;
        deferred = new Promise((r) => (resolve = r));
    }

    function passEvent() {
        // if no event to pass
        if (buffer.length === 0) {
            first = true; // reset first state
            return;
        }

        const eventsToEmit = [...buffer];
        const res = resolve;
        reset(false);
        setTimeout(passEvent, interval);
        res(eventsToEmit);
    }

    reset(true);
    destreamify(stream, (event) => {
        buffer.push(event);
        if (first) {
            passEvent();
        }
    });

    while (true) {
        // @ts-expect-error: deferred is definitly assigned before in `reset` call
        yield deferred;
    }
}

/**
 * @template T
 * @param {AsyncIterable<T>} stream
 * @param {(event: T) => void} callback
 */
async function destreamify(stream, callback) {
    for await (let event of stream) {
        callback(event);
    }
}

/**
 * @template T
 * @param {AsyncIterable<T>[]} iterable
 * @returns {AsyncIterable<T>}
 */
export async function* combine(iterable) {
    const asyncIterators = Array.from(iterable, (o) => o[Symbol.asyncIterator]());
    const results = [];
    let count = asyncIterators.length;
    const never = new Promise(() => {});

    /**
     * @param {AsyncIterator<T>} asyncIterator
     * @param {number} index
     * @returns
     */
    function getNext(asyncIterator, index) {
        return asyncIterator.next().then((result) => ({
            index,
            result,
        }));
    }

    const nextPromises = asyncIterators.map(getNext);

    try {
        while (count) {
            const { index, result } = await Promise.race(nextPromises);
            if (result.done) {
                nextPromises[index] = never;
                results[index] = result.value;
                count--;
            } else {
                nextPromises[index] = getNext(asyncIterators[index], index);
                yield result.value;
            }
        }
    } finally {
        for (const [index, iterator] of asyncIterators.entries())
            if (nextPromises[index] != never && iterator.return != null) {
                iterator.return();
            }
    }
    return results;
}
