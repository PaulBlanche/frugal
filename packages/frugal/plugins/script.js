export * from "../src/plugins/script/script.js";
