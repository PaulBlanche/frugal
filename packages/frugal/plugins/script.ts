export type * from "./script.d.ts";
export * from "./script.js";
