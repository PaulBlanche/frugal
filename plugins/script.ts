export { script } from '../src/build/plugins/script.ts';
