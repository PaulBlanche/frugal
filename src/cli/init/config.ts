import * as path from "../../../dep/std/path.ts";
import * as dotenv from "../../../dep/std/flags.ts";
import * as colors from "../../../dep/std/fmt/colors.ts";
import { CacheType, ExportType, Framework, toCacheType, toExportType, toFramework, VALID_CACHE } from "./types.ts";

export type Config = {
    root: string;
    exportType: ExportType;
    cacheType: CacheType;
    framework: Framework;
    force: boolean;
};

export function getConfig(args: string[]): Config {
    const flags = dotenv.parse(args, {
        string: ["export", "cache", "framework"],
        boolean: ["force", "help"],
        default: {
            "force": false,
        },
    });

    if (flags._[0]) {
        const exportType = toExportType(flags.export);
        if (exportType === undefined) {
            throw Error(
                `invalid export type "${flags.export}". Valid values are: ${Object.values(ExportType).join(", ")}`,
            );
        }

        const cacheType = toCacheType(flags.cache);
        if (cacheType === undefined || VALID_CACHE[exportType].includes(cacheType)) {
            throw Error(
                `invalid cache type "${flags.cache}". Valid values are: ${VALID_CACHE[exportType].join(", ")}`,
            );
        }

        const framework = toFramework(flags.framework);
        if (framework === undefined) {
            throw Error(
                `invalid cache type "${flags.framework}". Valid values are: ${Object.values(Framework).join(", ")}`,
            );
        }

        const root = path.resolve(String(flags._[0]));

        if (!isEmpty(root) && !flags.force) {
            console.error(`The directory ${root} is not empty. Use --force to overwrite it.`);
            Deno.exit(1);
        }

        return {
            root,
            exportType,
            cacheType,
            framework,
            force: flags.force,
        };
    }

    const name = askUntil("Enter a project name", {
        default: "frugal-project",
    });
    const root = path.resolve(name);

    const force = isEmpty(root) ||
        confirm(colors.yellow(`The directory ${root} is not empty. Do you want to overwrite it ?`));

    if (!isEmpty(root) && !force) {
        console.error(`The directory is not empty.`);
        Deno.exit(1);
    }

    const exportType = askUntil("Choose an export type", {
        head: "Projects have to be exported for specific plateform to be deployed.",
        default: ExportType.NO_EXPORT,
        values: Object.values(ExportType),
        sanitize: (exportType) => toExportType(exportType),
    });

    const cacheType = VALID_CACHE[exportType].length === 1
        ? VALID_CACHE[exportType][0]
        : askUntil("Choose an cache type", {
            head: "Frugal will have to cache static pages in some storage.",
            default: VALID_CACHE[exportType][0],
            values: VALID_CACHE[exportType],
            sanitize: (cacheType) => toCacheType(cacheType),
        });

    const framework = askUntil("Choose an framework", {
        head: "Frugal works with different frameworks.",
        default: Framework.NO_FRAMEWORK,
        values: Object.values(Framework),
        sanitize: (exportType) => toFramework(exportType),
    });

    return {
        root,
        exportType,
        cacheType,
        framework,
        force,
    };
}

function ask(message: string): string | undefined;
function ask(message: string, _default: string): string;
function ask(message: string, _default?: string): string | undefined;
function ask(message: string, _default?: string): string | undefined {
    const value = prompt(message, _default ? colors.gray(_default) : undefined);
    if (value === null) {
        return undefined;
    }
    return colors.stripColor(value);
}

type AskUntilConfig<VALUE> = {
    head?: string;
    default?: string;
    values?: string[];
    sanitize?: (userValue: string | undefined) => VALUE | undefined;
};

function askUntil<VALUE = string>(
    message: string,
    { head, default: _default, values, sanitize = (x) => (x as any) }: AskUntilConfig<VALUE>,
): VALUE {
    const askMessageParts = [colors.bold(message)];
    if (values) {
        askMessageParts.push(`(${values.join(", ")})`);
    }

    const askMessage = askMessageParts.join(" ");

    console.log("");
    head && console.log(colors.gray(`>> ${head}`));

    while (true) {
        const userValue = ask(askMessage, _default);

        const sanitizedValue = sanitize(userValue);

        if (sanitizedValue !== undefined) {
            return sanitizedValue;
        }

        const errorMessageParts = [`invalid value "${userValue}".`];
        if (values) {
            errorMessageParts.push(`Valid values are: ${values.join(", ")}`);
        }

        console.log(colors.red(errorMessageParts.join(" ")));
    }
}

function isEmpty(path: string) {
    try {
        const dir = [...Deno.readDirSync(path)];
        const isEmpty = dir.length === 0 ||
            dir.length === 1 && dir[0].name === ".git";

        return isEmpty;
    } catch (error) {
        if ((error instanceof Deno.errors.NotFound)) {
            return true;
        }
        throw error;
    }
}
