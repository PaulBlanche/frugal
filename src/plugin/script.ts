import * as esbuild from "../../dep/esbuild.ts";
import * as path from "../../dep/std/path.ts";
import * as fs from "../../dep/std/fs.ts";
import { denoLoaderPlugin, denoResolverPlugin } from "../../dep/esbuild_deno_loader.ts";

import { Plugin } from "../Plugin.ts";
import { log } from "../log.ts";
import { cleanOutdir } from "../build/plugins/cleanOutdir.ts";
import { outputMetafile } from "../build/plugins/outputMetafile.ts";
import { FrugalConfig } from "../Config.ts";

type ScriptOptions = {
    outdir: string;
    filter: RegExp;
};

type ContextCache = { id: string; context: esbuild.BuildContext };

export function script(
    { outdir = "js/", filter = /\.script.[tj]sx?$/ }: Partial<ScriptOptions> = {},
): Plugin {
    return (frugal) => {
        let contextCache: ContextCache | undefined = undefined;

        return {
            name: "frugal:script",
            setup(build) {
                build.onLoad({ filter }, async (args) => {
                    const url = frugal.url(args);
                    const contents = await frugal.load(url);
                    return { loader: "ts", contents };
                });

                build.onEnd(async (result) => {
                    const metafile = result.metafile;
                    const errors = result.errors;

                    if (errors.length !== 0 || metafile === undefined) {
                        return;
                    }

                    const assets = frugal.collect(filter, metafile);

                    const facadesMap: Record<string, Facade> = {};

                    for (const asset of assets) {
                        const entrypoint = asset.entrypoint;
                        const facadePath = path.resolve(frugal.config.tempdir, "asset", "script", entrypoint);
                        const facadeContent = `import "${asset.path}";`;
                        facadesMap[entrypoint] = facadesMap[entrypoint] ?? {
                            entrypoint,
                            path: facadePath,
                            content: [],
                        };
                        facadesMap[entrypoint].content.push(facadeContent);
                    }

                    const facades = Object.values(facadesMap);

                    await Promise.all(
                        facades.map(async (facade) => {
                            await fs.ensureFile(facade.path);
                            await Deno.writeTextFile(facade.path, facade.content.join("\n"));
                        }),
                    );

                    const context = await getContext(
                        facades.map((facade) => facade.path),
                        frugal.config,
                    );

                    const buildResult = await context.rebuild();

                    frugal.output("script", {
                        type: "page",
                        assets: getJsBundle(buildResult.metafile, facades, frugal.config),
                    });
                });

                async function getContext(
                    entrypoints: string[],
                    config: FrugalConfig,
                ): Promise<
                    esbuild.BuildContext<
                        Omit<esbuild.BuildOptions, "metafile"> & { metafile: true }
                    >
                > {
                    const id = entrypoints.slice().sort().join("");
                    if (contextCache?.id !== id) {
                        if (contextCache !== undefined) {
                            log(
                                "script entrypoints list has changed, recreate esbuild context",
                                { level: "debug", scope: "plugin:script", extra: entrypoints.join("\n") },
                            );

                            await contextCache.context.dispose();
                        }
                        contextCache = { id, context: await setupContext(entrypoints, config) };
                    }

                    return contextCache.context;
                }

                function setupContext(
                    entrypoints: string[],
                    config: FrugalConfig,
                ): Promise<
                    esbuild.BuildContext<
                        Omit<esbuild.BuildOptions, "metafile"> & { metafile: true }
                    >
                > {
                    const parentConfig = build.initialOptions;
                    const outdirPath = path.resolve(config.publicdir, outdir);
                    const esbuildConfig: esbuild.BuildOptions = {
                        ...config.esbuildOptions,
                        target: config.esbuildOptions?.target === undefined ||
                                config.esbuildOptions.target === "es6-modules"
                            ? ["es2020", "edge88", "firefox78", "chrome87", "safari14"]
                            : config.esbuildOptions.target,
                        define: {
                            ...config.esbuildOptions?.define,
                            "import.meta.environment": "'client'",
                        },
                        entryNames: "[dir]/[name]-[hash]",
                        chunkNames: "[dir]/[name]-[hash]",
                        assetNames: "[dir]/[name]-[hash]",
                        outdir: outdirPath,
                        plugins: [
                            denoResolverPlugin({
                                importMapURL: frugal.config.importMapURL?.href,
                            }),
                            ...(parentConfig.plugins?.filter((plugin) =>
                                plugin.name.startsWith("frugal:") && plugin.name !== "frugal:script"
                            ) ?? []),
                            ...(frugal.config.esbuildOptions?.plugins ?? []),
                            denoLoaderPlugin({
                                importMapURL: frugal.config.importMapURL?.href,
                                nodeModulesDir: true,
                            }),
                            cleanOutdir(config, false),
                            outputMetafile(),
                        ],
                        loader: { ".css": "empty" },
                        format: "esm",
                        entryPoints: entrypoints,
                        bundle: true,
                        metafile: true,
                        absWorkingDir: config.outdir,
                    };

                    log(`asset esbuild config`, {
                        level: "verbose",
                        scope: "plugin:script",
                        extra: JSON.stringify(esbuildConfig),
                    });

                    return esbuild.context(esbuildConfig);
                }
            },
        };
    };
}

function getJsBundle(
    metafile: esbuild.Metafile,
    facades: Facade[],
    config: FrugalConfig,
) {
    const generated: Record<string, string> = {};

    const outputs = Object.entries(metafile.outputs);

    for (const [outputPath, output] of outputs) {
        if (output.entryPoint === undefined) {
            continue;
        }

        const outputEntrypointPath = path.resolve(config.outdir, output.entryPoint);

        const facade = facades.find((facade) => facade.path === outputEntrypointPath);

        if (facade === undefined) {
            continue;
        }

        const jsBundleUrl = path.resolve(config.outdir, outputPath);
        const bundleName = path.relative(config.publicdir, jsBundleUrl);

        generated[facade.entrypoint] = `/${bundleName}`;
    }

    return generated;
}

type Facade = { entrypoint: string; path: string; content: string[] };
