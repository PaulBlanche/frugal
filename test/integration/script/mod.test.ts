import * as asserts from '../../../dep/std/testing/asserts.ts';
import { assertSnapshot } from '../../../dep/std/testing/snapshot.ts';
import { script } from '../../../src/build/plugins/script.ts';
import { getBuilder, withBrowser } from '../../_utils.ts';
import config from './frugal.config.ts';

Deno.test('script_loader: file structure', async (t) => {
  const builder = getBuilder({
    self: import.meta.url,
    pages: ['./page-bar.ts', './page-foo.ts'],
    log: { level: 'silent' },
    plugins: [
      script(),
    ],
  });

  const frugal = await builder.build();

  const responseCache = await frugal.config.responseCache('runtime');
  const pageFoo1 = (await responseCache.get('/foo/1'))?.body!;
  const pageFoo2 = (await responseCache.get('/foo/2'))?.body!;
  const pageBar1 = (await responseCache.get('/bar/1'))?.body!;
  const pageBar2 = (await responseCache.get('/bar/2'))?.body!;

  await assertSnapshot(t, pageFoo1);
  await assertSnapshot(t, pageFoo2);
  await assertSnapshot(t, pageBar1);
  await assertSnapshot(t, pageBar2);

  await builder.clean();
});

Deno.test('script: script execution and order', async () => {
  const builder = getBuilder(config);

  await builder.build();

  // trick deno check into no analysing dynamic import, because imported file
  // does not exists yet
  const serverScript = './cli/_server.ts';
  const { serve } = await import(serverScript);

  const controller = new AbortController();
  serve({ signal: controller.signal });

  await withBrowser(async (browser) => {
    const page = await browser.newPage();

    await page.goto('http://localhost:8000/foo/1');

    const _logFoo = await page.evaluate('_log');

    asserts.assertEquals(_logFoo, ['component', 'foo', 'shared']);

    await page.goto('http://localhost:8000/bar/1');

    const _logBar = await page.evaluate('_log');

    asserts.assertEquals(_logBar, ['component', 'shared', 'bar']);
  }, {
    onClose: () => controller.abort(),
  });

  await builder.clean();
});
