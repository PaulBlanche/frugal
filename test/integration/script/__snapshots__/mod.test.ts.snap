export const snapshot = {};

snapshot[`script_loader: file structure 1`] = `
'<html>
    <head>
        <title>foo</title>
    </head>
    <body>
        <div id="log"></div>
        <script type="module" src="/js/page-foo.js"></script>
    </body>
</html>'
`;

snapshot[`script_loader: file structure 2`] = `
'<html>
    <head>
        <title>foo</title>
    </head>
    <body>
        <div id="log"></div>
        <script type="module" src="/js/page-foo.js"></script>
    </body>
</html>'
`;

snapshot[`script_loader: file structure 3`] = `
'<html>
    <head>
        <title>bar</title>
    </head>
    <body>
        <div id="log"></div>
        <script type="module" src="/js/page-bar.js"></script>
    </body>
</html>'
`;

snapshot[`script_loader: file structure 4`] = `
'<html>
    <head>
        <title>bar</title>
    </head>
    <body>
        <div id="log"></div>
        <script type="module" src="/js/page-bar.js"></script>
    </body>
</html>'
`;
